### PR TITLE
improved structural equivalence checking

### PR DESCRIPTION
**Pull request by balazske**
Added check for `CXXMethodDecl` (not complete: some template cases missing?).
Cleaned up check of FunctionTemplateDecl.
Added unit tests.

Diagnostic messages not added.

### CODE REVIEW

**Comment by balazske** [comment]
The produced diagnostic messages depend on the order in source code of testing structural equivalence of elements. If a test for an element (for example field) fails the check is stopped and no other elements (for example methods) are checked. Probably it is better to check for everything and produce a message for every problem.

The check for fields in `RecordDecl` was moved to before check for methods, otherwise the check may fail at a method first and produce different message than before, this causes one unit test to fail.

File: unittests/AST/StructuralEquivalenceTest.cpp [hunk 1]
```diff
@@ -209,5 +209,165 @@ TEST_F(StructuralEquivalenceTest, WrongOrderOfFieldsInClass) {
...
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Delete) {
+  auto t = makeNamedDecls("struct foo { void x(); };",
+                          "struct foo { void x() = delete; };", Lang_CXX11);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Constructor) {
+  auto t = makeNamedDecls("struct foo { };",
+                          "struct foo { foo(); };", Lang_CXX);
```
Comment (martong) [inline unittests/AST/StructuralEquivalenceTest.cpp]: In this case it is not exactly clear which Decl is matched by the lookup inside `makeNamedDecls`. Is it the CXXRecordDecl or the CXXConstructorDecl? I think we could reuse `ASTImporterTestBase::getTuDecl()` here so we could explicitly match for the Decls we want to compare. (For that, we need a common TestBase class).

File: unittests/AST/StructuralEquivalenceTest.cpp [hunk 2]
```diff
@@ -209,5 +209,165 @@ TEST_F(StructuralEquivalenceTest, WrongOrderOfFieldsInClass) {
...
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Pure) {
+  auto t = makeNamedDecls("struct foo { virtual void x(); };",
+                          "struct foo { virtual void x() = 0; };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Const) {
+  auto t = makeNamedDecls("struct foo { void x(); };",
+                          "struct foo { void x() const; };", Lang_CXX);
```
Comment (martong) [inline unittests/AST/StructuralEquivalenceTest.cpp]: Could you please add tests for the `&&` qualifier too?
```
struct foo { void x() &&; };
```
Also, for `noexcept` and `throw()` functions?
```
void f0;
void f1() throw();
void f2() noexcept;
```
f0 should be structurally eq with f1, but not with f2, because noexcept is part of the type system.

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 3]
```diff
@@ -831,6 +831,67 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
   return true;
 }
 
+/// Determine structural equivalence of two methodss.
+static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
+                             CXXMethodDecl *Method1, CXXMethodDecl *Method2) {
+  if (Method1->isStatic() != Method2->isStatic())
```
Comment (martong) [inline lib/AST/ASTStructuralEquivalence.cpp]: Could we check for `&&` qualifier as well?
Is `noexcept` checked during normal function type check (in the call of line 888)?

**Review by martong — COMMENTED** [review]
Nice work! I especially like the tests. Still, I had some comments. Also, before merging, I think we should test this on bitcoin and protobuf with the custom jenkins job we have (I'll help you with that).

File: unittests/AST/StructuralEquivalenceTest.cpp (hunk 2 above)
Comment (martong) [reply unittests/AST/StructuralEquivalenceTest.cpp]: Probably we will have to add some tests in a disabled form (`noexcept` will fail).

File: unittests/AST/StructuralEquivalenceTest.cpp (hunk 1 above)
Comment (balazske) [reply unittests/AST/StructuralEquivalenceTest.cpp]: I this case it should be the struct foo because a lookup is performed on the top level DeclContext. There is an assertion that asserts if multiple results are found.

Comment (martong) [reply unittests/AST/StructuralEquivalenceTest.cpp]: OK.

File: unittests/AST/StructuralEquivalenceTest.cpp (hunk 2 above)
Comment (balazske) [reply unittests/AST/StructuralEquivalenceTest.cpp]: So there should be no check for exception specification in "throw" form at all?
But the `noexcept` should be checked, and then check the related expressions (if exist) too?

There is check for exceptions in `isStructurallyEquivalent` (QualType) that never checks anything because the "canonical type" does not contain exception information. (It is possible to use the original non-canonical type here.)

Comment (martong) [reply unittests/AST/StructuralEquivalenceTest.cpp]: `throw()` should be equal to `noexcept(true)` and with `noexcept`. Any other forms of throw we should ignore.

```
void f1() noexcept(expr1);
void f2() noexcept(expr2);
```
Above `f1` and `f2` should have the same type only if both expressions evaluates to the same value. But probably this would require constexpr evaluation which we might not be able to do at this phase.

Okay, I see now it is getting complicated... perhaps this should be kept as future work.

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 4]
```diff
@@ -831,6 +831,67 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
...
+  if (Method1->isVirtual() != Method2->isVirtual())
+    return false;
+  
+  if (Method1->isPure() != Method2->isPure())
+    return false;
+  if (Method1->isDeleted() != Method2->isDeleted())
+    return false;
+  // do not check for defaulted
+  
+  if (CXXConstructorDecl *Constructor1 = dyn_cast<CXXConstructorDecl>(Method1)) {
```
Comment (Xazax-hun) [inline lib/AST/ASTStructuralEquivalence.cpp]: Use `auto *` where type would otherwise be repeated.

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 5]
```diff
@@ -982,6 +1081,30 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
         }
         return false;
       }
+
+      // Check the methods for consistency.
+      CXXRecordDecl::method_iterator Method2 = D2CXX->method_begin(),
+                                     Method2End = D2CXX->method_end();
+      for (CXXRecordDecl::method_iterator Method1 = D1CXX->method_begin(),
```
Comment (Xazax-hun) [inline lib/AST/ASTStructuralEquivalence.cpp]: Can we actually ensure that the order of the methods is the same?
Right now the order of importing messes up everything. After import, we artificially reorder the fields.
Do we want to do the same for methods? It might add more runtime overhead, so I am unsure. Maybe we are better off leaving the order of methods arbitrary and not check for structural equivalence at all.

**Review by Xazax-hun — COMMENTED** [review]
Is this change supposed to fix something or we do this to catch more bugs? What is the runtime overhead? Maybe it would be better to only do these checks if some debug flags or macros are turned on?

File: lib/AST/ASTStructuralEquivalence.cpp (hunk 5 above)
Comment (balazske) [reply lib/AST/ASTStructuralEquivalence.cpp]: If order of methods is wrong the same problem would exist for fields and friends and maybe base classes too.
If the check is not done this can cause bad code imports.

Comment (Xazax-hun) [reply lib/AST/ASTStructuralEquivalence.cpp]: Yeah, but fields/friend are already reordered I think. We could always reorder more, but since right not it is a quadratic algorithm (due to linked lists being involved), we should keep the number of items to reorder minimal.

**Comment by balazske** [comment]
Code was updated.

File: lib/AST/ASTStructuralEquivalence.cpp (hunk 5 above)
Comment (balazske) [reply lib/AST/ASTStructuralEquivalence.cpp]: `ASTNodeImporter::ImportDeclContext` does a reorder for every `Decl` that should include fields, friends, methods. Anyway, in structural equivalence check, it would be better to sort the items in some way.

File: unittests/AST/StructuralEquivalenceTest.cpp [hunk 6]
```diff
@@ -209,5 +209,206 @@ TEST_F(StructuralEquivalenceTest, WrongOrderOfFieldsInClass) {
...
+TEST_F(StructuralEquivalenceTest, Noexcept3) {
+  auto t = makeNamedDecls("void foo() throw();",
+                          "void foo() noexcept(false);", Lang_CXX11);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceTest, Noexcept4) {
+  auto t = makeNamedDecls("void foo() throw();",
+                          "void foo() noexcept(true);", Lang_CXX11);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
```
Comment (martong) [inline unittests/AST/StructuralEquivalenceTest.cpp]: This should be `EXPECT_TRUE` since `noexcept(true)` equals to `noexcept`.
Rather than `EXPECT_FALSE` we should make the test `DISABLED_`.

**Review by martong — COMMENTED** [review]
Nice work and good tests.

**Comment by martong** [comment]
Merged this branch on my repo on top of `fix_structural_eq_cache_problem_by_fixing_the_recursion` and started a custom jenkins build on that, so we can see the effects.

File: unittests/AST/StructuralEquivalenceTest.cpp (hunk 6 above)
Comment (balazske) [reply unittests/AST/StructuralEquivalenceTest.cpp]: `// yes: the expression is not evaluated at check`
I wanted to notice that the current implementation does not consider the given expression at all. If the expression is omitted (`noexcept`) it is equivalent to `throw()` but not if there is any expression (including `true`).

ASTStructuralEquivalence.cpp line 441 checks for `noexcept` and `throw()`:
```
      if ((Spec1 != EST_DynamicNone || Spec2 != EST_BasicNoexcept) &&
          (Spec1 != EST_BasicNoexcept || Spec2 != EST_DynamicNone) &&
           Spec1 != Spec2)
```
I did not want to make additional check for ComputedNoexcept and test for a `true` expression or any other that evaluates to `true`.

**Comment by martong** [comment]
Balazs, I think we should cut this patch into two:

1) All those improvements which are connected to the equivalence check of free functions and member functions (noexcept, && qualifier, const qualifier, etc). This includes the new function
```
+static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
+                             CXXMethodDecl *Method1, CXXMethodDecl *Method2) {
```
and it's tests too.

2) Checking the member functions as well in a RecordDecl.

I think 1) is an immediate improvement and @Xazax-hun would accept that as well. Just think about out-of-class definition of member functions ... it is definitely an improvement that we can distinguish a cons qualified member function from a non-const one. 

2) might cause some degradation since our general structural eq checking is still buggy (and also we create sometimes a faulty AST with respect to classtemplates and specializations). Thus, perhaps 2) should be merged after we are done with #384 / 1)

**Comment by Xazax-hun** [comment]
I support martong's proposal!

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 7]
```diff
@@ -1505,6 +1587,14 @@ bool StructuralEquivalenceContext::Finish() {
         // Kind mismatch.
         Equivalent = false;
       }
+    } else if (CXXMethodDecl *MD1 = dyn_cast<CXXMethodDecl>(D1)) {
```
Comment (martong) [inline lib/AST/ASTStructuralEquivalence.cpp]: :+1:

File: unittests/AST/ASTImporterTest.cpp [hunk 8]
```diff
@@ -2689,6 +2689,157 @@ TEST_P(ASTImporterTestBase, ImportDefinitionOfClassTemplateAfterFwdDecl) {
...
+    auto *FromD = FirstDeclMatcher<CXXRecordDecl>().match(
+        FromTU, cxxRecordDecl(hasName("A")));
+
+    ToD2 = Import(FromD, Lang_CXX);
+  }
+  
+  EXPECT_NE(ToD1, ToD2);
+}
+
+TEST_P(ASTImporterTestBase, DISABLED_ImportOfNonEquivalentRecordWithFunction) {
```
Comment (martong) [inline unittests/AST/ASTImporterTest.cpp]: I think it would worth to add an other similar test,  where we import the functionDecl(hasName('x')) and not the cxxRecordDecl itself. In that case structural eq will dispatch to comparing the methods, so `EXPECT_EQ(ToD1, ToD2)` will be successful.

File: unittests/AST/StructuralEquivalenceTest.cpp [hunk 9]
```diff
@@ -209,5 +201,219 @@ TEST_F(StructuralEquivalenceTest, WrongOrderOfFieldsInClass) {
...
+TEST_F(StructuralEquivalenceTest, ParamPtr) {
+  auto t = makeNamedDecls("void foo(int *);",
+                          "void foo(int);", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+struct StructuralEquivalenceCXXMethodTest : StructuralEquivalenceTest {
+};
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Virtual) {
```
Comment (martong) [inline unittests/AST/StructuralEquivalenceTest.cpp]: I think one out-of-class definition equivalence test where we compare two out-of-class definitions would be useful.
Something like:

```
TEST_F(StructuralEquivalenceCXXMethodTest, OutOfClass) {
  auto t = makeDecls<FunctionDecl>(
      "struct X { virtual void f(); }; void X::f() { }",
      "struct X { void f(); }; void X::f() { }",
      Lang_CXX,
      functionDecl(allOf(hasName("f"), isDefinition())));
  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
}
```

**Review by martong — COMMENTED** [review]
This looks very good, still I think we have to add a few more test cases.

File: unittests/AST/ASTImporterTest.cpp (hunk 8 above)
Comment (balazske) [reply unittests/AST/ASTImporterTest.cpp]: Now there is a ImportOf{Equivalent|NonEquivalent}RecordWith{Field|Function} test (4 tests). Is it OK to duplicate every of them with a version that imports the member `x` instead of the record `A`? (And there is the possibility to add the case with friends.) The import of `x` can be inserted into the existing tests too.

Comment (martong) [reply unittests/AST/ASTImporterTest.cpp]: > The import of x can be inserted into the existing tests too.

Perhaps that would be the best.

File: unittests/AST/StructuralEquivalenceTest.cpp [hunk 10]
```diff
@@ -209,5 +201,316 @@ TEST_F(StructuralEquivalenceTest, WrongOrderOfFieldsInClass) {
...
+  EXPECT_TRUE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceTest, CompareSameDeclWithMultiple) {
+  auto t = makeNamedDecls(
+      "struct A{}; struct B{}; void foo(A a, A b);",
+      "struct A{}; struct B{}; void foo(A a, B b);",
+      Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
```
Comment (balazske) [inline unittests/AST/StructuralEquivalenceTest.cpp]: This test is not strictly related but useful for something.

**Comment by balazske** [comment]
Maybe `final` should be checked too?

**Comment by martong** [comment]
With respect to functions, if we check purity then it seems logical to check finality too.

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 11]
```diff
@@ -411,8 +414,11 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
   }
 
   case Type::FunctionProto: {
-    const FunctionProtoType *Proto1 = cast<FunctionProtoType>(T1);
-    const FunctionProtoType *Proto2 = cast<FunctionProtoType>(T2);
+    const auto *Proto1 = cast<FunctionProtoType>(T1);
+    const auto *Proto2 = cast<FunctionProtoType>(T2);
+    const FunctionProtoType *OrigProto1 = cast<FunctionProtoType>(OrigT1);
+    const FunctionProtoType *OrigProto2 = cast<FunctionProtoType>(OrigT2);
```
Comment (Xazax-hun) [inline lib/AST/ASTStructuralEquivalence.cpp]: We could use auto here like above.

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 12]
```diff
@@ -422,23 +428,36 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
...
-    if (Proto1->getExceptionSpecType() == EST_Dynamic) {
-      if (Proto1->getNumExceptions() != Proto2->getNumExceptions())
+    
+    // Check exceptions: This information may get lost in canonical type.
+    auto Spec1 = OrigProto1->getExceptionSpecType();
+    auto Spec2 = OrigProto2->getExceptionSpecType();
+    
+    if (!isUnresolvedExceptionSpec(Spec1) &&
+        !isUnresolvedExceptionSpec(Spec2)) {
+      if ((Spec1 != EST_DynamicNone || Spec2 != EST_BasicNoexcept) &&
```
Comment (Xazax-hun) [inline lib/AST/ASTStructuralEquivalence.cpp]: Why do we need this complex check here? Isn't `Spec1 != Spec2` sufficient?

Comment (Xazax-hun) [reply lib/AST/ASTStructuralEquivalence.cpp]: Also, exception specifications only part of the type system after C++17 (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/p0012r1.html). Maybe we should not require the exception specifications to be the same unless C++17 or above is used?

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 13]
```diff
@@ -831,6 +850,69 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
   return true;
 }
 
+/// Determine structural equivalence of two methodss.
+static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
+                             CXXMethodDecl *Method1, CXXMethodDecl *Method2) {
```
Comment (Xazax-hun) [inline lib/AST/ASTStructuralEquivalence.cpp]: The formatting is off here, use clang format.

File: lib/AST/ASTStructuralEquivalence.cpp [hunk 14]
```diff
@@ -831,6 +850,69 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
...
+    return false;
+  if (Method1->isVolatile() != Method2->isVolatile())
+    return false;
+  if (Method1->isVirtual() != Method2->isVirtual())
+    return false;
+  if (Method1->isPure() != Method2->isPure())
+    return false;
+  if (Method1->isDeleted() != Method2->isDeleted())
+    return false;
+  // do not check for defaulted
```
Comment (Xazax-hun) [inline lib/AST/ASTStructuralEquivalence.cpp]: Comments should be full sentences. Also, why don't we check for that?

File: lib/AST/ASTStructuralEquivalence.cpp (hunk 7 above)
Comment (Xazax-hun) [reply lib/AST/ASTStructuralEquivalence.cpp]: Type is repeated, use auto.

**Review by Xazax-hun — DISMISSED** [review]

File: lib/AST/ASTStructuralEquivalence.cpp (hunk 12 above)
Comment (martong) [reply lib/AST/ASTStructuralEquivalence.cpp]: If we have `foo () noexcept` in one TU and `foo ()` in another TU, is that an ODR violation?
Or that violates ODR only in C++17? I think it is hard to answer.

File: lib/AST/ASTStructuralEquivalence.cpp (hunk 13 above)
Comment (balazske) [reply lib/AST/ASTStructuralEquivalence.cpp]: Reformat the whole file (it is full of such format problems)?

Comment (balazske) [reply lib/AST/ASTStructuralEquivalence.cpp]: Replace every redundant typename with `auto` (can be done in lot of places in the file)?

Comment (Xazax-hun) [reply lib/AST/ASTStructuralEquivalence.cpp]: I think only reformat/replace in the code you change. If we are about to clean up old code, it is better done in a separate pull request.

**Comment by balazske** [comment]
There is a class named `ODRHash`, is it not useful for these checks?

**Comment by martong** [comment]
@balazske 
Yes, I think we try to implement something similar here. So if `ODRHash` is stable and full we could use that instead of the whole messy `ASTStructuralEquivalence`.  Maybe in the future we can/should change to that.

File: lib/AST/ASTStructuralEquivalence.cpp (hunk 12 above)
Comment (balazske) [reply lib/AST/ASTStructuralEquivalence.cpp]: For now: The "structurally equivalent" check is not an exact ODR check (it should check function body too and other things). It is to determine if two declarations "look similar". (I can not find "C11 6.2.7/1" to check what "structurally compatible layout" means here, `Sema::hasStructuralCompatLayout` is the single non-ASTImporter use.)

**Comment by balazske** [comment]
Code was updated, the new warnings about incompatible methods are still not implemented.

**Review by martong — APPROVED** [review]
LGTM! :)

**Comment by balazske** [comment]
Is it possible to merge this into ctu-clang6 on local computer and push it to ericsson repo (without pull request)? (I do not want to use "web editor" and there exists the branch on local repository that would be easy to merge. But this would make 2 commits (the existing and the merge) and probably other not wanted things. Otherwise the web editor is usable but no compile is done after it to check for basic syntax errors.)

**Comment by martong** [comment]
@balazske 
Why do you want to push it to ericsson repo? Do you want it to be in a separate branch in the ericsson repo? You can test your changes even if they are kept in your remote repo (github/balazske/clang).

Here is how you can do that.
You can merge the tip of Ericsson/ctu-clang6 branch into you local branch. Then push that to a branch in your remote repo.
Then you can modify the jenkins job http://cc.elte.hu:15009/job/test-clang-with-projects-egbomrt/ to use your repo and branch.

**Comment by martong** [comment]
Started a dry run in CI:
http://cc.elte.hu:15009/job/test-clang-with-projects-egbomrt/9/

Expect the results in 9 hours.

**Comment by Xazax-hun** [comment]
In bitcoin the failed actions increased from 20 to 45. We should check if we caught more bugs with the checks or there are bugs in the checks themselves.

**Comment by martong** [comment]
Yes.
Balazs, could you please check why we get so many new assertions:
```
Assertion `isa(Val) && "cast() argument of incompatible type!"' failed. [33]
```
http://cc.elte.hu:15009/job/test-clang-with-projects-egbomrt/9/HTML_Report/

**Comment by balazske** [comment]
How to find the log files (or other information) with this assertion? (Is there filename and line number, and call stack?)

**Comment by martong** [comment]
I think the best is to reproduce locally, on you machine the assert.
You should use the latest CodeChecker configured to use your built clang and then a ctu-analyze.
```
cd bitcoin
./configure
CodeChecker analyze --analyzers clangsa --ctu-collect -o ./reports ./compile_commands.json -j4
CodeChecker analyze --analyzers clangsa --ctu-collect -o ./reports ./compile_commands.json -j4
```
Use these setting to configure bitcoin:
```
    {
      "name": "Bitcoin",
      "url": "https://github.com/bitcoin/bitcoin.git",
      "tag": "v0.16.0",
      "configure_command": "./autogen.sh && ./configure --disable-wallet --disable-static --disable-tests"
    },

```

**Comment by martong** [comment]
Note, bitcoin CTU run is fast (20 minutes)

**Comment by martong** [comment]
@balazske 
Is your last commit fixes the extra assertions?

**Comment by martong** [comment]
Just started a new build:
http://cc.elte.hu:15009/job/test-clang-with-projects-egbomrt/10

**Comment by balazske** [comment]
It fixes the cast error.
There is an `ToDecl->hasBody()` assertion that is not fixed. But the fix for that is easy (`setEncounteredUnsupportedNode` is not called), not committed yet. (The `CXXStdInitializerListExpr` is not handled at import.)

**Comment by martong** [comment]
Yes, please do a different pull request for the initializer list.

**Comment by balazske** [comment]
Import of std initializer list: #391 .

**Comment by balazske** [comment]
LLVM review:
https://reviews.llvm.org/D48628

**Comment by balazske** [comment]
Backport from LLVM: #425
